openApiSync/openApiSync.go

```diff
@@ -0,0 +1,205 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package openApiSync
+
+import (
+	"bytes"
+	"fmt"
+	"go/ast"
+	"go/printer"
+	"go/token"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/getkin/kin-openapi/openapi3"
+	"github.com/pkg/errors"
+	"github.com/sajari/fuzzy"
+	"golang.org/x/tools/go/analysis"
+)
+
+var Analyzer = &analysis.Analyzer{
+	Name: "openApiSync",
+	Doc:  "check for inconsistencies between OpenAPI spec and the source code",
+	Run:  run,
+}
+
+var specFile string
+
+func init() {
+	Analyzer.Flags.StringVar(&specFile, "spec", "", "Path to the OpenAPI 3 YAML spec file")
+}
+
+func formatNode(fset *token.FileSet, node interface{}) string {
+	var typeNameBuf bytes.Buffer
+	printer.Fprint(&typeNameBuf, fset, node)
+	return typeNameBuf.String()
+}
+
+func processRouterInit(pass *analysis.Pass, name string, routerPrefixes map[string]string, swagger *openapi3.Swagger, cm *fuzzy.Model) {
+	for _, file := range pass.Files {
+		ast.Inspect(file, func(n ast.Node) bool {
+			decl, ok := n.(*ast.FuncDecl)
+			if !ok || decl.Name.Name != name {
+				return true
+			}
+			for _, stmt := range decl.Body.List {
+				expr, ok := stmt.(*ast.ExprStmt)
+				if !ok {
+					continue
+				}
+				aexpr, ok := expr.X.(*ast.CallExpr).Fun.(*ast.SelectorExpr).X.(*ast.CallExpr)
```

Comment:
This looks a bit fragile, but because is really tided to our source code I don't see it as a big deal.